pkg/reconciler/pipelinerun/resources/apply.go

```diff
@@ -264,31 +286,58 @@ func ApplyTaskResultsToPipelineResults(
 					case v1beta1.ParamTypeArray:
 						if stringIdx != "*" {
 							intIdx, _ := strconv.Atoi(stringIdx)
-							stringReplacements[variable] = resultValue.ArrayVal[intIdx]
+							if intIdx < len(resultValue.ArrayVal) {
+								stringReplacements[variable] = resultValue.ArrayVal[intIdx]
+							} else {
+								invalidPipelineResults = append(invalidPipelineResults, pipelineResult.Name)
+								validPipelineResult = false
+							}
 						} else {
 							arrayReplacements[v1beta1.StripStarVarSubExpression(variable)] = resultValue.ArrayVal
 						}
+					case v1beta1.ParamTypeObject:
+						objectReplacements[v1beta1.StripStarVarSubExpression(variable)] = resultValue.ObjectVal
```

Comment:
as part of #5001 it would be good to move this out of v1beta1. `StripStarVarSubExpression` doesn't seem like a function we'd want to export from that package.

Reply:
Yeah sure, I will check and fix this

Reply:
move this function to substitution